File: app/itinerary/planner.py
"""
Trip itinerary planner.

Assembles TripCandidate objects from the pre-computed route-stations index,
user inputs (departure coordinates, days, rhythm), and the Eurovelo route data.
Does not make any network calls — pure data assembly.
"""

import json
import math
from dataclasses import dataclass, field

from app.constants import (
    OUTBOUND_CANDIDATE_COUNT,
    ROUTE_START_ZONE_FRACTION,
    ROUTE_START_ZONE_MAX_KM,
)
from app.geo.distance import haversine_km
from app.geo.station_matcher import StationOnRoute
from app.itinerary.rhythm import get_rhythm, total_biking_km


@dataclass
class TripCandidate:
    """
    A candidate bike + train itinerary along one Eurovelo route.

    Attributes:
        route_id: Eurovelo route identifier (e.g. 'EV15').
        route_name: Human-readable route name.
        departure_station: Station the user arrives at by train (outbound).
        arrival_station: Station the user departs from by train (return).
        biking_start_km: Cumulative km on the route where biking begins.
        biking_end_km: Cumulative km on the route where biking ends.
        total_biking_km: Total distance cycled (km).
        n_days: Number of days for the whole trip.
        rhythm_key: Rhythm key used to compute the biking distance.
        geometry: Full-resolution list of [lat, lon] pairs for the biked segment
                  (for map rendering), sliced from the route's track_points
                  between the start and end stations.
    """

    route_id: str
    route_name: str
    departure_station: StationOnRoute
    arrival_station: StationOnRoute
    biking_start_km: float
    biking_end_km: float
    total_biking_km: float
    n_days: int
    rhythm_key: str
    geometry: list[list[float]] = field(default_factory=list)


def load_route_index(index_path: str) -> dict:
    """
    Load and validate the route_stations.json proximity index.

    Args:
        index_path: Path to the JSON file produced by scripts/preprocess.py.

    Returns:
        Parsed dict with at least 'routes' and 'generated_at' keys.

    Raises:
        FileNotFoundError: If the file does not exist.
        ValueError: If the file is not valid JSON or missing required keys.
    """
    with open(index_path, encoding="utf-8") as f:
        data = json.load(f)
    if "routes" not in data:
        raise ValueError("Index file is missing 'routes' key")
    return data


def _deserialize_station(station_dict: dict) -> StationOnRoute:
    """
    Convert a station dict from the JSON index into a StationOnRoute dataclass.

    Args:
        station_dict: Dict with keys nom, libellecourt, codes_uic, lat, lon,
                      distance_to_route_km, cumulative_km.

    Returns:
        StationOnRoute dataclass instance.
    """
    return StationOnRoute(
        nom=station_dict["nom"],
        libellecourt=station_dict["libellecourt"],
        codes_uic=station_dict["codes_uic"],
        lat=station_dict["lat"],
        lon=station_dict["lon"],
        distance_to_route_km=station_dict["distance_to_route_km"],
        cumulative_km=station_dict["cumulative_km"],
    )


def get_stations_near_route_start(
    route_data: dict,
    departure_lat: float,
    departure_lon: float,
    n_candidates: int,
) -> list[StationOnRoute]:
    """
    Select candidate outbound stations from the start zone of a route.

    The start zone is the first ROUTE_START_ZONE_FRACTION of the route, capped at
    ROUTE_START_ZONE_MAX_KM. Among stations in the start zone, the n_candidates
    closest to the departure city (by haversine) are returned.

    Args:
        route_data: Single route dict from the index (key 'stations', 'total_km').
        departure_lat: Latitude of the user's departure city, decimal degrees.
        departure_lon: Longitude of the user's departure city, decimal degrees.
        n_candidates: Maximum number of stations to return.

    Returns:
        List of StationOnRoute sorted by ascending distance to departure city.
        May be shorter than n_candidates if fewer stations exist in the start zone.
    """
    total_km = route_data["total_km"]
    start_zone_km = min(
        total_km * ROUTE_START_ZONE_FRACTION,
        ROUTE_START_ZONE_MAX_KM,
    )
    stations_in_zone = [
        _deserialize_station(s)
        for s in route_data["stations"]
        if s["cumulative_km"] <= start_zone_km
    ]
    stations_in_zone.sort(
        key=lambda s: haversine_km(departure_lat, departure_lon, s.lat, s.lon)
    )
    return stations_in_zone[:n_candidates]


def compute_end_station(
    route_data: dict,
    start_station: StationOnRoute,
    biking_km: float,
) -> StationOnRoute | None:
    """
    Find the station closest to the expected end point after biking.

    The end point is start_station.cumulative_km + biking_km along the route.
    The station whose cumulative_km is nearest to this target is returned.

    Args:
        route_data: Single route dict from the index.
        start_station: Station where biking begins (outbound arrival).
        biking_km: Total biking distance in km.

    Returns:
        The StationOnRoute closest to the end point, or None if no stations
        exist in the index for this route.
    """
    stations = route_data["stations"]
    if not stations:
        return None
    target_km = start_station.cumulative_km + biking_km
    best = min(stations, key=lambda s: abs(s["cumulative_km"] - target_km))
    return _deserialize_station(best)


def downsample_geometry(
    points: list[tuple[float, float]],
    max_points: int,
) -> list[list[float]]:
    """
    Downsample a list of (lat, lon) tuples to at most max_points entries.

    Selects every N-th point where N = ceil(len(points) / max_points).
    Always includes the first and last point.

    Args:
        points: Full list of (lat, lon) tuples.
        max_points: Maximum number of points to return.

    Returns:
        List of [lat, lon] lists (JSON-serializable). Empty if points is empty.
    """
    if not points:
        return []
    if len(points) <= max_points:
        return [[lat, lon] for lat, lon in points]
    step = math.ceil(len(points) / max_points)
    sampled = [points[i] for i in range(0, len(points), step)]
    if sampled[-1] != points[-1]:
        sampled.append(points[-1])
    return [[lat, lon] for lat, lon in sampled]


def round_geometry(
    points: list[tuple[float, float]],
    decimals: int,
) -> list[list[float]]:
    """
    Return every point as a [lat, lon] list with coordinates rounded.

    Unlike downsample_geometry, NO vertices are dropped — the full GPX
    resolution is preserved so the drawn line follows the real itinerary.
    Coordinates are only rounded to `decimals` places to reduce JSON size
    (5 dp ≈ 1.1 m, well below visible map precision).

    Args:
        points: Full list of (lat, lon) tuples.
        decimals: Number of decimal places to keep for each coordinate.

    Returns:
        List of [lat, lon] lists (JSON-serializable). Empty if points is empty.
    """
    if not points:
        return []
    return [[round(lat, decimals), round(lon, decimals)] for lat, lon in points]


def _nearest_point_index(
    polyline: list[tuple[float, float]],
    lat: float,
    lon: float,
) -> int:
    """
    Find the index of the polyline vertex closest to a geographic point.

    Linear argmin of the haversine distance over every vertex; used to snap a
    station to the nearest point on the route polyline.

    Args:
        polyline: Ordered list of (lat, lon) tuples.
        lat: Target point latitude in decimal degrees.
        lon: Target point longitude in decimal degrees.

    Returns:
        Index of the nearest vertex, or -1 if the polyline is empty.
    """
    if not polyline:
        return -1
    best_idx = 0
    best_dist = math.inf
    for i, (p_lat, p_lon) in enumerate(polyline):
        dist = haversine_km(lat, lon, p_lat, p_lon)
        if dist < best_dist:
            best_dist = dist
            best_idx = i
    return best_idx


def _extract_segment_points(
    route_data: dict,
    start_station: StationOnRoute,
    end_station: StationOnRoute,
) -> list[tuple[float, float]]:
    """
    Extract the route track points between two stations.

    Snaps each station to the nearest vertex of the route's 'track_points'
    polyline and returns the inclusive slice between those two indices. Slicing
    the same full-resolution polyline that is drawn guarantees the segment ends
    land on the route line next to the station markers, regardless of station
    spacing. The result is full-resolution (no downsampling).

    Args:
        route_data: Single route dict from the index (with 'track_points').
        start_station: Bike departure station (carries lat/lon).
        end_station: Bike arrival station (carries lat/lon).

    Returns:
        List of (lat, lon) tuples for the segment. Empty if track_points is
        absent.
    """
    raw_points = route_data.get("track_points")
    if not raw_points:
        return []
    polyline = [(p[0], p[1]) for p in raw_points]
    i1 = _nearest_point_index(polyline, start_station.lat, start_station.lon)
    i2 = _nearest_point_index(polyline, end_station.lat, end_station.lon)
    if i1 == -1 or i2 == -1:
        return []
    return polyline[min(i1, i2): max(i1, i2) + 1]


def find_itinerary_candidates(
    route_id: str,
    route_data: dict,
    departure_lat: float,
    departure_lon: float,
    n_days: int,
    rhythm_key: str,
) -> list[TripCandidate]:
    """
    Build TripCandidate list for a single Eurovelo route.

    For each outbound candidate station (up to OUTBOUND_CANDIDATE_COUNT):
    1. Compute the end station based on total biking km.
    2. Build the biked-segment geometry by slicing the route's full-resolution
       track_points between the start and end stations' nearest vertices
       (see _extract_segment_points), so the line spans station to station.
    3. Assemble a TripCandidate.

    Args:
        route_id: Eurovelo route identifier.
        route_data: Route dict from the proximity index.
        departure_lat: Latitude of the user's departure city.
        departure_lon: Longitude of the user's departure city.
        n_days: Total trip days.
        rhythm_key: Rhythm key string.

    Returns:
        List of TripCandidate objects. May be empty if no start stations exist.
    """
    rhythm = get_rhythm(rhythm_key)
    biking_km = total_biking_km(n_days, rhythm)
    route_name = route_data.get("name", route_id)

    outbound_candidates = get_stations_near_route_start(
        route_data, departure_lat, departure_lon, OUTBOUND_CANDIDATE_COUNT
    )
    if not outbound_candidates:
        return []

    candidates: list[TripCandidate] = []
    for start_station in outbound_candidates:
        end_station = compute_end_station(route_data, start_station, biking_km)
        if end_station is None:
            continue
        start_km = start_station.cumulative_km
        end_km = start_km + biking_km
        geometry = _extract_segment_points(route_data, start_station, end_station)
        candidates.append(
            TripCandidate(
                route_id=route_id,
                route_name=route_name,
                departure_station=start_station,
                arrival_station=end_station,
                biking_start_km=round(start_km, 1),
                biking_end_km=round(end_km, 1),
                total_biking_km=round(biking_km, 1),
                n_days=n_days,
                rhythm_key=rhythm_key,
                geometry=geometry,
            )
        )
    return candidates


def find_all_itineraries(
    route_ids: list[str],
    index: dict,
    departure_lat: float,
    departure_lon: float,
    n_days: int,
    rhythm_key: str,
) -> list[TripCandidate]:
    """
    Build TripCandidate lists for all requested Eurovelo routes.

    When a single route is requested, returns up to OUTBOUND_CANDIDATE_COUNT
    candidates. When multiple routes are requested, returns at most one
    candidate per route (the closest start station to the departure city).

    Args:
        route_ids: List of Eurovelo route identifiers to search.
        index: Full proximity index dict (from load_route_index()).
        departure_lat: Latitude of the user's departure city.
        departure_lon: Longitude of the user's departure city.
        n_days: Total trip days.
        rhythm_key: Rhythm key string.

    Returns:
        Flat list of TripCandidate objects across all routes. Empty if no
        candidates found.
    """
    routes = index.get("routes", {})
    multiple_routes = len(route_ids) > 1
    all_candidates: list[TripCandidate] = []

    for route_id in route_ids:
        if route_id not in routes:
            continue
        candidates = find_itinerary_candidates(
            route_id=route_id,
            route_data=routes[route_id],
            departure_lat=departure_lat,
            departure_lon=departure_lon,
            n_days=n_days,
            rhythm_key=rhythm_key,
        )
        if multiple_routes and candidates:
            # One result per route when multiple routes are selected
            all_candidates.append(candidates[0])
        else:
            all_candidates.extend(candidates)

    return all_candidates

File: static/js/planner.js
/**
 * planner.js — Pure JS itinerary planner (port of app/itinerary/).
 *
 * Computes bike trip candidates from the route_stations.json index without
 * any network calls. All constants match the Python values in app/constants.py.
 *
 * Exposes: window.InterPlanner
 */

(function () {
  "use strict";

  // ── Constants (must match app/constants.py) ───────────────────────────────

  /** @type {Object.<string, {speed_kmh: number, hours_per_day: number}>} */
  const RHYTHMS = {
    escargot:   { speed_kmh: 12.0, hours_per_day: 5.0 },
    randonneur: { speed_kmh: 15.0, hours_per_day: 6.5 },
    athlete:    { speed_kmh: 20.0, hours_per_day: 8.0 },
  };

  const ROUTE_START_ZONE_FRACTION = 0.15;
  const ROUTE_START_ZONE_MAX_KM   = 100.0;
  const OUTBOUND_CANDIDATE_COUNT  = 3;
  const EARTH_RADIUS_KM           = 6371.0;
  const HALF_DAY_FRACTION         = 0.5;

  // ── Geometry helpers ──────────────────────────────────────────────────────

  /**
   * Compute the great-circle distance between two WGS84 points (haversine).
   *
   * @param {number} lat1 - Latitude of point A in decimal degrees.
   * @param {number} lon1 - Longitude of point A in decimal degrees.
   * @param {number} lat2 - Latitude of point B in decimal degrees.
   * @param {number} lon2 - Longitude of point B in decimal degrees.
   * @returns {number} Distance in kilometres (non-negative).
   */
  function haversineKm(lat1, lon1, lat2, lon2) {
    const toRad = (d) => (d * Math.PI) / 180;
    const dLat = toRad(lat2 - lat1);
    const dLon = toRad(lon2 - lon1);
    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(toRad(lat1)) *
        Math.cos(toRad(lat2)) *
        Math.sin(dLon / 2) *
        Math.sin(dLon / 2);
    return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(a));
  }

  /**
   * Find the index of the polyline vertex closest to a geographic point.
   *
   * Linear argmin of the haversine distance over every vertex. Used to snap a
   * station to the nearest point on the route polyline.
   *
   * @param {Array<[number, number]>} polyline - Array of [lat, lon] pairs.
   * @param {number} lat - Target point latitude in decimal degrees.
   * @param {number} lon - Target point longitude in decimal degrees.
   * @returns {number} Index of the nearest vertex, or -1 if the polyline is empty.
   */
  function nearestPointIndex(polyline, lat, lon) {
    if (!polyline || polyline.length === 0) return -1;
    let bestIdx = 0;
    let bestDist = Infinity;
    for (let i = 0; i < polyline.length; i++) {
      const d = haversineKm(lat, lon, polyline[i][0], polyline[i][1]);
      if (d < bestDist) {
        bestDist = d;
        bestIdx = i;
      }
    }
    return bestIdx;
  }

  // ── Rhythm helpers ────────────────────────────────────────────────────────

  /**
   * Return the maximum biking distance in one full day for a rhythm key.
   *
   * @param {string} rhythmKey - One of 'escargot', 'randonneur', 'athlete'.
   * @returns {number} Distance in km.
   * @throws {Error} If rhythmKey is not recognised.
   */
  function kmPerFullDay(rhythmKey) {
    const r = RHYTHMS[rhythmKey];
    if (!r) throw new Error(`Unknown rhythm key: ${rhythmKey}`);
    return r.speed_kmh * r.hours_per_day;
  }

  /**
   * Compute total biking distance for a trip.
   *
   * - n_days == 1: half a day of biking (both trains on same day).
   * - n_days >= 2: (n_days - 1) full days (one day equivalent lost to trains).
   *
   * @param {number} nDays - Total trip days (1–15).
   * @param {string} rhythmKey - Rhythm key string.
   * @returns {number} Total biking km.
   */
  function totalBikingKm(nDays, rhythmKey) {
    const daily = kmPerFullDay(rhythmKey);
    if (nDays === 1) return HALF_DAY_FRACTION * daily;
    return (nDays - 1) * daily;
  }

  // ── Station selection ─────────────────────────────────────────────────────

  /**
   * Select candidate outbound stations from the start zone of a route.
   *
   * The start zone is the first ROUTE_START_ZONE_FRACTION of the route,
   * capped at ROUTE_START_ZONE_MAX_KM. Among those stations, the n closest
   * to the departure city (by haversine) are returned.
   *
   * @param {Object} routeData - Single route object from the route_stations index.
   * @param {number} depLat - Departure city latitude.
   * @param {number} depLon - Departure city longitude.
   * @param {number} n - Maximum number of stations to return.
   * @returns {Object[]} Station objects sorted by ascending distance to departure.
   */
  function getStationsNearRouteStart(routeData, depLat, depLon, n) {
    const totalKm = routeData.total_km;
    const zoneKm = Math.min(totalKm * ROUTE_START_ZONE_FRACTION, ROUTE_START_ZONE_MAX_KM);
    const inZone = routeData.stations.filter((s) => s.cumulative_km <= zoneKm);
    inZone.sort((a, b) => haversineKm(depLat, depLon, a.lat, a.lon) - haversineKm(depLat, depLon, b.lat, b.lon));
    return inZone.slice(0, n);
  }

  /**
   * Find the station closest to the expected end point after biking.
   *
   * The target position is startStation.cumulative_km + bikingKm.
   *
   * @param {Object} routeData - Single route object from the index.
   * @param {Object} startStation - Station where biking begins.
   * @param {number} bikingKm - Total biking distance in km.
   * @returns {Object|null} Station object closest to the end point, or null if
   *   no stations exist.
   */
  function computeEndStation(routeData, startStation, bikingKm) {
    const stations = routeData.stations;
    if (!stations || stations.length === 0) return null;
    const targetKm = startStation.cumulative_km + bikingKm;
    let best = stations[0];
    let bestDiff = Math.abs(best.cumulative_km - targetKm);
    for (let i = 1; i < stations.length; i++) {
      const diff = Math.abs(stations[i].cumulative_km - targetKm);
      if (diff < bestDiff) {
        bestDiff = diff;
        best = stations[i];
      }
    }
    return best;
  }

  /**
   * Extract the route track points between two stations.
   *
   * Snaps each station to the nearest vertex of the route's `track_points`
   * polyline and returns the (inclusive) slice between those two indices. Slicing
   * the same full-resolution polyline that is drawn guarantees the segment ends
   * land exactly on the route line next to the station markers, regardless of
   * station spacing. The result is full-resolution (no downsampling) so the bike
   * leg traces the GPX exactly.
   *
   * @param {Array<[number, number]>} trackPoints - Route polyline ([lat, lon]),
   *   ordered along the route. Empty/absent yields an empty result.
   * @param {{lat:number, lon:number}} startStation - Bike departure station.
   * @param {{lat:number, lon:number}} endStation - Bike arrival station.
   * @returns {Array<[number, number]>} [lat, lon] pairs for the biked segment,
   *   or [] if trackPoints is empty.
   */
  function extractSegmentPoints(trackPoints, startStation, endStation) {
    if (!trackPoints || trackPoints.length === 0) return [];
    const i1 = nearestPointIndex(trackPoints, startStation.lat, startStation.lon);
    const i2 = nearestPointIndex(trackPoints, endStation.lat, endStation.lon);
    if (i1 === -1 || i2 === -1) return [];
    return trackPoints.slice(Math.min(i1, i2), Math.max(i1, i2) + 1);
  }

  // ── Candidate assembly ────────────────────────────────────────────────────

  /**
   * Build TripCandidate objects for a single Eurovelo route.
   *
   * For each outbound candidate station (up to OUTBOUND_CANDIDATE_COUNT):
   *
   * 1. Biking budget — `totalBikingKm(nDays, rhythmKey)` converts the trip
   *    length and pace into a distance: a half day for a 1-day trip, otherwise
   *    `(nDays - 1)` full days, each full day being `speed_kmh * hours_per_day`
   *    for the chosen rhythm (escargot / randonneur / athlète).
   * 2. End station — `computeEndStation` scans the route's stations for the one
   *    whose `cumulative_km` is closest to `startStation.cumulative_km + bikingKm`,
   *    so the biked segment begins and ends at real reachable stations.
   * 3. Segment geometry — `extractSegmentPoints` snaps the start and end stations
   *    to their nearest vertices on the route's full-resolution `track_points`
   *    and returns the slice between them. The geometry is kept full-resolution
   *    so the bike leg traces the GPX exactly and its ends land on the station
   *    markers. (`startKm`/`endKm` are still used for the distance labels.)
   * 4. Hand-off — `search.js.buildItineraryCard` copies `candidate.geometry`
   *    onto the itinerary as `itinerary.geometry`, which map.js draws.
   *
   * @param {string} routeId - Eurovelo route ID (e.g. 'EV6').
   * @param {Object} routeData - Route entry from the proximity index.
   * @param {number} depLat - Departure city latitude.
   * @param {number} depLon - Departure city longitude.
   * @param {number} nDays - Total trip days.
   * @param {string} rhythmKey - Rhythm key string.
   * @returns {Object[]} Array of candidate objects. May be empty.
   */
  function findItineraryCandidates(routeId, routeData, depLat, depLon, nDays, rhythmKey) {
    const bikingKm = totalBikingKm(nDays, rhythmKey);
    const routeName = routeData.name || routeId;
    const outboundStations = getStationsNearRouteStart(routeData, depLat, depLon, OUTBOUND_CANDIDATE_COUNT);
    if (outboundStations.length === 0) return [];

    return outboundStations.reduce((acc, startStation) => {
      const endStation = computeEndStation(routeData, startStation, bikingKm);
      if (!endStation) return acc;
      const startKm = startStation.cumulative_km;
      const endKm = startKm + bikingKm;
      const geometry = extractSegmentPoints(routeData.track_points, startStation, endStation);
      acc.push({
        route_id: routeId,
        route_name: routeName,
        departure_station: startStation,
        arrival_station: endStation,
        biking_start_km: Math.round(startKm * 10) / 10,
        biking_end_km: Math.round(endKm * 10) / 10,
        total_biking_km: Math.round(bikingKm * 10) / 10,
        n_days: nDays,
        rhythm_key: rhythmKey,
        geometry: geometry,
      });
      return acc;
    }, []);
  }

  /**
   * Build candidates for all requested Eurovelo routes.
   *
   * When a single route is selected: up to OUTBOUND_CANDIDATE_COUNT results.
   * When multiple routes are selected: at most one result per route.
   *
   * @param {string[]} routeIds - List of Eurovelo route IDs to search.
   * @param {Object} index - Full proximity index (parsed route_stations.json).
   * @param {number} depLat - Departure city latitude.
   * @param {number} depLon - Departure city longitude.
   * @param {number} nDays - Total trip days.
   * @param {string} rhythmKey - Rhythm key string.
   * @returns {Object[]} Flat array of candidate objects.
   */
  function findAllItineraries(routeIds, index, depLat, depLon, nDays, rhythmKey) {
    const routes = (index && index.routes) || {};
    const multipleRoutes = routeIds.length > 1;
    const all = [];

    for (const routeId of routeIds) {
      if (!routes[routeId]) continue;
      const candidates = findItineraryCandidates(routeId, routes[routeId], depLat, depLon, nDays, rhythmKey);
      if (multipleRoutes && candidates.length > 0) {
        all.push(candidates[0]);
      } else {
        all.push(...candidates);
      }
    }
    return all;
  }

  // ── Public API ─────────────────────────────────────────────────────────────

  window.InterPlanner = {
    haversineKm,
    kmPerFullDay,
    totalBikingKm,
    getStationsNearRouteStart,
    computeEndStation,
    nearestPointIndex,
    extractSegmentPoints,
    findItineraryCandidates,
    findAllItineraries,
  };
})();
